Deprecate REST access to System Indices

AthenaEryma opened the pull request:
This PR adds deprecation warnings when accessing System Indices via the REST layer. At this time, these warnings are only enabled for Snapshot builds by default, to allow projects external to Elasticsearch additional time to adjust their access patterns.

Deprecation warnings will be triggered by all REST requests which access registered System Indices, except for purpose-specific APIs which access System Indices as an implementation detail and a few specific APIs which will continue to allow access to system indices by default:

- `GET _cluster/health`
- `GET {index}/_recovery`
- `GET _cluster/allocation/explain`
- `GET _cluster/state`
- `POST _cluster/reroute`
- `GET {index}/_stats`
- `GET {index}/_segments`
- `GET {index}/_shard_stores`
- `GET _cat/[indices,aliases,health,recovery,shards,segments]`

Deprecation warnings for accessing system indices take the form:
```
this request accesses system indices: [.some_system_index], but in a future major version, direct access to system indices will be prevented by default
```

AthenaEryma committed:
first cut of header-based system index enforcement

AthenaEryma committed:
Properly shut down components in test

AthenaEryma committed:
Version guard system index flag on cleanup

AthenaEryma committed:
Fix guard in REST layer

AthenaEryma committed:
Set origin from Enrich

AthenaEryma committed:
Set origin from Enrich in the processor

AthenaEryma committed:
Fix NPEs in Enrich tests

AthenaEryma committed:
Ensure we never double-add the header (should be tests only?)

AthenaEryma committed:
Fix NPE in UsageServiceTests

AthenaEryma committed:
Fix NPE in RestAnalyzeActionTests

AthenaEryma committed:
Fix NPE in RestBulkActionTests

AthenaEryma committed:
Fix NPE/mocking in `RestActionTestCase`s

AthenaEryma committed:
Add `allow_system_index_access` flag to search spec

AthenaEryma committed:
Add `allow_system_index_access` as necessary to Watcher tests

AthenaEryma committed:
Add `allow_system_index_access` flag to count, index, index.delete specs

AthenaEryma committed:
Add `allow_system_index_access` to ML REST tests as necessary

AthenaEryma committed:
Allow REST handlers to control whether they allow system index access

AthenaEryma committed:
Fix NPE/client mock in RestSearchScrollActionTests

AthenaEryma committed:
Fix NPE/client mock in RestClearScrollActionTests

AthenaEryma committed:
Add allow_system_index_access flag to Get API

AthenaEryma committed:
Add allow_system_index_access flag to Security tests as necessary

AthenaEryma committed:
Add allow_system_index_access flag to upgrade tests as necessary

AthenaEryma committed:
Include system indices in cluster health by default

AthenaEryma committed:
Fix NPE/client mock in SecurityBaseRestHandlerTests

AthenaEryma committed:
Fix NPE/client mock in RestHasPrivilegesActionTests

AthenaEryma committed:
Add allow_system_index_access flag to ML Restart tests as necessary

AthenaEryma committed:
Add allow_system_index_access flag to cluster restart tests as necessary

AthenaEryma committed:
Add allow_system_index_access flag to Transform tests as necessary

AthenaEryma committed:
Clean up old comments

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Allow Kibana APIs to access system indices by default.

AthenaEryma committed:
Fix NPE/client mock in RestGetUserPrivilegesActionTests

AthenaEryma committed:
Rename and centralize system index access control header key

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Add allow_system_index_access flag to Transform Upgrade tests as necessary

jaymode commented on server/src/main/java/org/elasticsearch/cluster/ClusterModule.java:
```diff
@@ -107,13 +108,13 @@
 
     public ClusterModule(Settings settings, ClusterService clusterService, List<ClusterPlugin> clusterPlugins,
-                         ClusterInfoService clusterInfoService) {
+                         ClusterInfoService clusterInfoService, ThreadPool threadPool) {
```
since we only need the `ThreadContext` here can we pass that as the constructor argument instead of the `ThreadPool`?

jaymode commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstraction.java:
```diff
@@ -210,6 +217,13 @@ public boolean isHidden() {
+        public boolean isSystem() {
+            //G-> This is probably not the best way to tell if an alias is a system-index-alias
+            // this should probably be checked on the alias layer and added as a property in the AliasMetadata
+            return referenceIndexMetadatas.stream().anyMatch(IndexMetadata::isSystem);
```
I agree and think we should make this similar to what we did with hidden aliases

jaymode commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -61,17 +62,25 @@
 public class IndexNameExpressionResolver {
 
     public static final String EXCLUDED_DATA_STREAMS_KEY = "es.excluded_ds";
+    public static final String SYSTEM_INDEX_ACCESS_CONTROL_KEY = "_prevent_system_index_access";
```
this is a nit, I personally like `_system_index_access_allowed`

jaymode commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -650,6 +665,13 @@ boolean isPatternMatchingAllIndices(Metadata metadata, String[] indicesOrAliases
     }
 
+    private boolean isSystemIndexAccessAllowed() {
+        if (threadContext == null) {
```
why would the threadContext be null?

gwbrown replied:
I think this was a hack I put in during development to work around some (since otherwise fixed) test issues that accidentally got committed. I'll remove it, thanks for the catch.

jaymode commented on server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java:
```diff
@@ -76,6 +79,16 @@ public final long getUsageCount() {
+            if (threadContext.getHeader(SYSTEM_INDEX_ACCESS_CONTROL_KEY) == null
+                && (allow_system_index_access == null
+                || allow_system_index_access.isEmpty()
+                || Boolean.parseBoolean(allow_system_index_access) == false)) {
```
Please use `Booleans.parseBoolean` instead, see #59190. This will also remove the need for the null/empty checks.

jaymode commented on server/src/test/java/org/elasticsearch/action/support/AutoCreateIndexTests.java:
```diff
@@ -177,7 +178,8 @@ public void testUpdate() {
         ClusterSettings clusterSettings = new ClusterSettings(settings,
                 ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
-        AutoCreateIndex  autoCreateIndex = new AutoCreateIndex(settings, clusterSettings, new IndexNameExpressionResolver());
+        AutoCreateIndex  autoCreateIndex = new AutoCreateIndex(settings, clusterSettings,
```
```suggestion
        AutoCreateIndex autoCreateIndex = new AutoCreateIndex(settings, clusterSettings,
```

jaymode commented on server/src/test/java/org/elasticsearch/rest/action/document/RestBulkActionTests.java:
```diff
@@ -45,32 +46,38 @@
-                    ).withMethod(RestRequest.Method.POST).build(),
-                mock(RestChannel.class), mockClient
-            );
-        Mockito.verify(mockClient)
```
By removing this, we no longer verify that the bulk method is actually called. I think adding an AtomicInteger as a counter for method invocations and validating that the method is called once should be sufficient

gwbrown replied:
++, I did that for a number of other tests that needed similar change, but forgot on this one.

jaymode commented on test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:
```diff
@@ -642,9 +642,13 @@ private void wipeCluster() throws Exception {
 
     protected static void wipeAllIndices() throws IOException {
         boolean includeHidden = minimumNodeVersion().onOrAfter(Version.V_7_7_0);
+        boolean includeSystem = minimumNodeVersion().onOrAfter(Version.V_8_0_0);
```
Do you want to have a constant version somewhere for when the `allow_system_index_access` flag was added?

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlIndexAndAlias.java:
```diff
@@ -99,7 +99,7 @@ public static void createIndexAndAliasIfNecessary(Client client,
         String firstConcreteIndex = indexPatternPrefix + "-000001";
         String[] concreteIndexNames =
-            resolver.concreteIndexNames(clusterState, IndicesOptions.lenientExpandOpen(), indexPattern);
+            resolver.concreteIndexNames(clusterState, IndicesOptions.lenientExpandOpen(), indexPattern); //G-> Does this work as expected?
```
I guess this needs to be added to the TODO list

jaymode reviewed (commented):
Overall I think this looks good. In terms of restricting which system indices can be accessed by a given plugin, did you plan to implement that here or as a follow-up?

AthenaEryma committed:
Add allow_system_index_access flag to relevant REST specs

AthenaEryma committed:
ClusterModule now requires a ThreadContext instead of a ThreadPool

AthenaEryma committed:
Remove unnecessary null check in IndexNameExpressionResolver

AthenaEryma committed:
s/Boolean.parseBoolean/Booleans.parseBoolean

AthenaEryma committed:
Remove accidental double space per review

AthenaEryma committed:
Ensure method is actually called in RestBulkActionTests

AthenaEryma committed:
Move allow_system key to constant & add version constant

AthenaEryma committed:
Refactor ThreadContext header to use positive logic

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Fix test after inverting header logic

AthenaEryma committed:
Fix logic when attaching header

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Fix compilation from merge

AthenaEryma committed:
Add `allow_system_index_access` flag to test

AthenaEryma committed:
Remove unnecessary comment

AthenaEryma committed:
INER tests for total wildcards

AthenaEryma committed:
Add dedicated system index blocking/access integration test

AthenaEryma committed:
Rules for aliases & system indices + unit tests

AthenaEryma committed:
Clean up system index upgrade tests

AthenaEryma committed:
Add allow_system_index_access flag to docs as necessary & docs updates

AthenaEryma committed:
Add allow_no_indices to upgrade tests to ensure they fail if appropriate

AthenaEryma committed:
Ensure we throw a 404 if a system index is explicitly requested, but not allowed

AthenaEryma committed:
Upgrade tests for alias logic

gwbrown commented:
I believe I've addressed all of the outstanding review comments, although I've deviated somewhat from some of the things discussed:
- Docs are pretty slim, I have an appointment to talk to some docs folks tomorrow.
- System Aliases aren't like hidden aliases, as that added a lot of overhead. This implementation allows an alias to refer to either only system indices, or regular indices, but not both. There is an exception for system indices created in 7.0, with unit & upgrade tests to ensure this works properly.
- I haven't yet converted this to issue deprecation warnings instead of just enforcing access restrictions. I still intend to, but that will likely be tomorrow, in a fairly small commit that should be independently reviewable.

gwbrown commented:
I'll also update the PR description before I convert this from a draft.

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Fix BWC tests to account for both upgrades from 7.x and current

jaymode commented on test/framework/src/main/java/org/elasticsearch/test/client/NoOpNodeClient.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+public class NoOpNodeClient extends NodeClient {
```
can you add basic javadocs?

jaymode commented on test/framework/src/main/java/org/elasticsearch/test/rest/RestActionTestCase.java:
```diff
@@ -66,4 +78,54 @@ protected void dispatchRequest(RestRequest request) {
         controller.dispatchRequest(request, channel, threadContext);
     }
+
+    public static class VerifyingClient extends NoOpNodeClient {
```
can you add basic class level javadocs?

jaymode commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/TransportExecuteEnrichPolicyAction.java:
```diff
@@ -58,7 +61,7 @@ public TransportExecuteEnrichPolicyAction(
             settings,
             clusterService,
-            client,
+            new OriginSettingClient(client, ENRICH_ORIGIN),
```
I don't know that this is correct. If I am not mistaken this is to take a user's policy and execute it to create the enrich index for the policy. The user's permissions dictate what indices they can access so we do not want to override that with the enrich origin when reading the data from the source index but instead need to do it only when writing the data to the enrich system indices unless I missed something?

gwbrown replied:
Looking at this closer, I think you're correct - I think I misread what this does the first time. Thanks for the catch!

(Also, do we not have enrich+security tests to catch something like this? I'll have to check)

jaymode replied:
It looks like we do have some security tests for Enrich (`x-pack/plugin/enrich/qa/rest-with-security`) but it is a limited set of tests.

gwbrown replied:
We discussed this offline and came to the conclusion that this is indeed the incorrect way to handle this. I've updated this to ensure that all actions that touch the `source` index in an Enrich policy run in the user's security context, and the Enrich security context is used only for operations that only touch the `.enrich*` index.

This does run into a problem in the reindex call used to populate the `.enrich*` index, which touches both. In this case, we simply drop the response headers added during the reindex call - a change which will need a follow-up later, as that approach will not work when we enable full protection of system indices.

jaymode commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/AbstractEnrichProcessor.java:
```diff
@@ -188,8 +191,9 @@ int getMaxMatches() {
+        Client originClient = new OriginSettingClient(client, ENRICH_ORIGIN);
         return (req, handler) -> {
-            client.execute(
+            originClient.execute(
```
I think this one is right since this is used to access the enrich system indices.

gwbrown replied:
++, this one I'm pretty confident on being the right thing.

jaymode reviewed (dismissed):
Looking good overall but I left some comments

gwbrown commented on docs/reference/rest-api/common-parms.asciidoc:
```diff
@@ -228,6 +228,14 @@ Wildcard expressions are not accepted.
+(Optional, boolean) If `true`, allows this request to access system indices
+directly. This option should be used with caution, as system indices are not
+intended to be accessed directly.
+Defaults to `false`.
```
You're right, it's not false everywhere - some APIs (like Cluster Health) override it:
https://github.com/gwbrown/elasticsearch/blob/8d4f90c9c6f8dc0a35b2bbfbc6dd6f348a7cf1da/server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java#L55-L58

I'll update the docs.

gwbrown commented on qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SystemIndicesUpgradeIT.java:
```diff
@@ -23,73 +23,85 @@
+                "  ]\n" +
+                "}");
+            assertThat(client().performRequest(putAliasRequest).getStatusLine().getStatusCode(), is(200));
+        } else if (CLUSTER_TYPE == ClusterType.UPGRADED) {
```
That won't work, I think, because there's also a `ClusterType.MIXED` that this might be called with.

jaymode replied:
You're right; I missed that the line with `ClusterType.MIXED` was a deletion only.

gwbrown commented on server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java:
```diff
@@ -53,6 +58,8 @@
     public static final Setting<Boolean> MULTI_ALLOW_EXPLICIT_INDEX =
         Setting.boolSetting("rest.action.multi.allow_explicit_index", true, Property.NodeScope);
+    public static final Version ALLOW_SYSTEM_INDEX_ADDED_VERSION = Version.V_8_0_0;
+    public static final String ALLOW_SYSTEM_INDEX_ACCESS_KEY = "allow_system_index_access";
```
This string is the REST parameter, as in `GET _search?allow_system_index_access=true`. The constant defining the key used for the `ThreadContext` header is defined in [IndexNameExpressionResolver](https://github.com/elastic/elasticsearch/blob/a17155c1de71005e39be572c36816a6a63fbe047/server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java#L67).

My general philosophy has been to put the constant definitions as close to their "primary consumer" as possible.

gwbrown replied:
That said, this tells me that these constants probably need names that are clearer about their purpose. So I'll update the names to make them clearer.

jaymode replied:
Thanks; maybe `ALLOW_SYSTEM_INDEX_ACCESS_PARAM`?

gwbrown commented on server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java:
```diff
@@ -76,6 +83,14 @@ public final long getUsageCount() {
+        if (allowSystemIndexAccessByDefault() == false) {
+            final String allowSystemIndexParameter = request.param(ALLOW_SYSTEM_INDEX_ACCESS_KEY);
+            final ThreadContext threadContext = client.threadPool().getThreadContext();
+            if (threadContext.getHeader(SYSTEM_INDEX_ACCESS_CONTROL_KEY) == null
```
👍 Yeah, this is primarily due to wrapping of handlers, and actually IIRC it only triggers in a few test cases that do somewhat odd things. I'll look into moving it up a level, thanks for the suggestion.

AthenaEryma committed:
Rename constants to be clearer

AthenaEryma committed:
Move constants & default access config to `RestHandler` (from `BaseRestHandler`)

AthenaEryma committed:
Move header attachment to `RestController#dispatchRequest` (from `BaseRestHandler`)

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Use the correct ThreadContext in RestController

AthenaEryma committed:
Javadoc per review

AthenaEryma committed:
Ensure allowSystemIndexAccessByDefault is passed through SecurityRestFilter

AthenaEryma committed:
Switch to emitting a deprecation warning instead of changing behavior

AthenaEryma committed:
Merge branch 'master' into si/protection-header

gwbrown commented:
There's still one known test failure in `TransformSurvivesUpgradeIT`, which is down to a call that needs to have the parameter `allow_system_index_access` if talking to >=8.0, but NOT have it if talking to <8.0, in a mixed cluster test where it is impossible to know the version of the node that will receive the request.

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Correct NoOpNodeClient for changes in NodeClient

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Implement "feature flag" for choosing system index access behavior

AthenaEryma committed:
Docs changes per review

AthenaEryma committed:
Formatting changes per review

AthenaEryma committed:
allow_system_index_access flag in YAML tests as necessary

AthenaEryma committed:
Allow system index access in GetAliases, as disallowing it causes very odd behavior

AthenaEryma committed:
Allow system index access in GetDataStreamsStats, as disallowing it causes very odd behavior

AthenaEryma committed:
Allow system index access in cat APIs, otherwise we get deprecation warnings

AthenaEryma committed:
Remove allow_system flag from test that no longer needs it due to changed defaults

AthenaEryma committed:
Remove allow_system flag from more tests that no longer needs it due to changed defaults

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Add allow_system_index_access to docs tests where necessary

Note: these should be removed and replaced with actual APIs everywhere this is necessary.

AthenaEryma committed:
Fix origin setting in Enrich classes

AthenaEryma committed:
Add allow_system_index_access to reindex tests as necessary

AthenaEryma committed:
Add allow_system_index_access to Watcher tests & test cleanup as necessary

AthenaEryma committed:
Add allow_system_index_access to Monitoring test cleanup as necessary

AthenaEryma committed:
Add allow_system_index_access to Transform test cleanup as necessary

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
add allow_system_index_access to Watcher cleanup

AthenaEryma committed:
Remove now unnecessary allow_system_index_access from upgrade test

AthenaEryma committed:
Add allow_system_index_access to ML tests as necessary

AthenaEryma committed:
Another allow_system_index_access test fix

AthenaEryma committed:
Remove allow_system_index_access from TransformSurvivesUpgradeIT

AthenaEryma committed:
Add allow_system_index_access as necessary in MlMappingsUpgradeIT

AthenaEryma committed:
Ensure system index access is allowed for GetAliasesAction even if there's a header to disallow it

Otherwise, we get unnecessary deprecation warnings

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Drop response headers during SQL index resolution to prevent deprecation warnings

AthenaEryma committed:
Drop response headers during Enrich reindex to prevent deprecation warnings

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Fix compilation error due to merge

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
ThreadPool -> ThreadContext in test per review

AthenaEryma committed:
Formatting per review

AthenaEryma committed:
Add comment on enrich hack

gwbrown commented:
@elasticmachine test this please

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Expect deprecation warnings in ML upgrade tests

AthenaEryma committed:
Merge branch 'master' into si/protection-header

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Core)

pugnascotia commented on docs/reference/api-conventions.asciidoc:
```diff
@@ -42,6 +42,19 @@ include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=expand-wildcards]
 The defaults settings for the above parameters depend on the API being used.
 
+Some indices (hereafter "system indices") are used by various system
+components and/or plugins to store state or configuration. These indices
```
Do we define what a "system component" is?

gwbrown replied:
This should probably be `modules and/or plugins`, thanks.

pugnascotia commented on qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java:
```diff
@@ -1386,67 +1390,97 @@ public void testResize() throws Exception {
-        assertBusy(() -> {
+    @SuppressWarnings("unchecked")
+    public void testSystemIndexMetadataIsUpgraded() throws Exception {
+        if (isRunningAgainstOldCluster()) {
```
Why have we combined two tests into one, where before they were using `assumeTrue` / `assumeFalse`?

gwbrown replied:
The code is logically the same, but combines the setup and check steps into one test method. 

Using `assumeTrue`/`assumeFalse` causes the assuming test method to effectively be skipped when the assumed condition is not met - and in this case, they weren't really separate test cases, but a setup step (on the old cluster) and a check step (on the new cluster). By combining them into one test method that branches based on which cluster it's running against, it makes it much easier to run the entire test, as you can now use `-Dtests.method=testSystemIndexMetadataIsUpgraded` instead of having to specify multiple methods.

pugnascotia commented on qa/smoke-test-http/src/test/java/org/elasticsearch/http/SystemIndexRestIT.java:
```diff
@@ -0,0 +1,172 @@
+            .build();
+        Request searchRequest = new Request("GET", "/" + queryPattern + randomFrom("/_count", "/_search"));
+        searchRequest.setJsonEntity("{\"query\": {\"match\":  {\"some_field\":  \"some_value\"}}}");
+        // Disallow no indices to cause an exception if this resolves to zero indices (as we expect)
```
I think this comment is incorrect?

gwbrown replied:
Yes, this is from an old version of the test, before I switched this to emit a deprecation warning instead of omitting system indices.

pugnascotia commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -282,10 +317,44 @@
+                    case RESTRICTED:
+                        final String errorMsg = "while resolving indices, resolved system indices [" + resolvedSystemIndices +
+                            "] when access to system indices is restricted";
+                        assert false : errorMsg;
```
Should this be removed?

gwbrown replied:
This is intentional. This will cause the node to fail if this line is hit with assertions enabled (i.e. during testing), which is what I want as this would indicate there's a bug in the code which should cause system indices to be completely omitted - and if there is, I want it to yell as loudly as possible.

If this line is hit without assertions enabled, it will fall through to the next line, which throws an exception. This is a relatively common pattern to enforce invariants more strictly during testing than at runtime.

pugnascotia commented:
@gwbrown Can you comment on the changes around supplying an empty `ThreadContext`?

pugnascotia commented on server/src/test/java/org/elasticsearch/cluster/metadata/IndexAbstractionTests.java:
```diff
@@ -116,13 +118,82 @@ public void testHiddenAliasValidation() {
     }
 
-    private IndexMetadata buildIndexWithAlias(String indexName, String aliasName, @Nullable Boolean aliasIsHidden) {
+    public void testSystemAliasValidation() {
```
I'm not generally a fan of long tests that check lots of cases. Can we usefully split up this method into a series of more targeted test cases, ideally with a Javadoc comment that explains what is the test is for ([see point 9 here](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#the-long-version)).

gwbrown replied:
I've split this test and the one in `IndexNameExpressionResolverTests` into multiple tests. I have not added Javadoc to the split test cases though, as the test cases ended up fairly small and I was finding in attempting to write Javadoc for them that the Javadoc largely was just repeating the test names.

pugnascotia commented on server/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java:
```diff
@@ -1810,6 +1828,125 @@ public void testIgnoreThrottled() {
         }
     }
 
+    public void testSystemIndexResolutionWhenAllowed() {
```
Same observation here about splitting up this test.

pugnascotia commented on server/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java:
```diff
@@ -1810,6 +1828,125 @@ public void testIgnoreThrottled() {
+        {
+            SearchRequest request = new SearchRequest(".ml-meta");
+
+            List<String> indexNames = Arrays
+                .stream(indexNameExpressionResolver.concreteIndices(state, request))
+                .map(i -> i.getName())
+                .collect(Collectors.toList());
```
What do you think about defining a helper method, to cut down on the ceremony here? Something like:

```java
private void getConcreteIndexNames(ClusterState state, SearchRequest request) {
    return List<String> indexNames = Arrays
        .stream(indexNameExpressionResolver.concreteIndices(state, request))
        .map(i -> i.getName())
        .collect(Collectors.toList());
}
```

pugnascotia commented on x-pack/qa/src/main/java/org/elasticsearch/xpack/test/rest/IndexMappingTemplateAsserter.java:
```diff
@@ -51,9 +51,10 @@
      * @throws IOException On error
      */
-    public static void assertMlMappingsMatchTemplates(RestClient client) throws IOException {
+    public static void assertMlMappingsMatchTemplates(RestClient client, boolean allowSystemIndexWarnings) throws IOException {
```
I can't see this new parameter being used. Am I missing something?

gwbrown replied:
Good catch, I forgot some plumbing here. I'll fix - if this is `false`, it should then pass `false` to all of the `assertLegacyTemplateMatchesIndexMappings` calls (as opposed to now, where some of them are hardcoded to `true`)

gwbrown replied:
Actually, it turns out I misread one of the callsites and this must be true for both places this is called, so I removed this parameter and went back to specifying the values passed to `assertLegacyTemplateMatchesIndexMappings` directly.

AthenaEryma committed:
Wording change per review

AthenaEryma committed:
Update comment per review

gwbrown commented:
> @gwbrown Can you comment on the changes around supplying an empty `ThreadContext`?

Do you mean comment on this PR, or a comment in the code itself?

In production code, there's only a single `ThreadContext` object that manages `ThreadLocal`s to handle each thread's context, which will be passed in at node startup. `IndexNameExpressionResolver` now requires information from the the `ThreadContext`, so it's needed in the constructor, but if the test does not depend on varying behavior based on the `ThreadContext`, we can just use the default one.

AthenaEryma committed:
Factor out shared stream/map code per review

AthenaEryma committed:
Break up large INER test cases per review

AthenaEryma committed:
Tweak method name

AthenaEryma committed:
Actually pass parameter in ML upgrade tests per review

AthenaEryma committed:
Break up IndexAbstractionTests per review

AthenaEryma committed:
Merge branch 'master' into si/protection-header

gwbrown commented:
@pugnascotia Thanks for the review, I believe I've addressed all your comments - can you confirm?

gwbrown commented:
Something blew up trying to clone the repo for the Windows packaging sample.

@elasticmachine elasticsearch-ci/packaging-sample-windows

gwbrown commented:
Something blew up trying to clone the repo for the Windows packaging sample.

@elasticmachine run elasticsearch-ci/packaging-sample-windows

AthenaEryma committed:
Add origin for Logstash management APIs

AthenaEryma committed:
Always allow system indiex warnings in ML index template tests

jaymode commented on server/build.gradle:
```diff
@@ -298,3 +298,10 @@ tasks.named("licenseHeaders").configure {
+  /* We need to set this to prevent release builds from failing while the default is different between snapshot and release builds.
+  * In particular, this is needed for IndexNameExpressionResolverTests, which assumes the value is `deprecated`.
+  */
+  systemProperty 'es.system_index_access_behavior', 'deprecated'
```
I am not sure about this; shouldn't this be an `assume` in the tests?

gwbrown replied:
That sounds like a better solution - I just didn't think of it when I noticed the situation would happen. Will do that instead.

gwbrown replied:
I've removed the system property in these gradle files and used `assumeTrue` in the tests in question.

jaymode commented on qa/smoke-test-http/build.gradle:
```diff
@@ -34,4 +34,9 @@ integTest {
+  /* We need to set this here to prevent release builds from failing while the default is different between snapshot and release builds.
+  * Specifically, this is needed for SystemIndexRestIT, which assumes the value is `deprecated`.
+  */
+  systemProperty 'es.system_index_access_behavior', 'deprecated'
```
I am not sure about this; shouldn't this be an `assume` in the tests?

gwbrown replied:
I've removed the system property in these gradle files and used assumeTrue in the tests in question.

jaymode commented on server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java:
```diff
@@ -72,8 +72,15 @@ public String getName() {
+    @Override
+    public boolean allowSystemIndexAccessByDefault() {
+        // If not given an index (only an alias), resolves `_all`, which causes deprecation warnings.
+        return true;
```
Should resolution of _all that includes system indices trigger a deprecation warning? If we get all aliases and return system aliases such as `.security` in 7.x but stop doing so in 8.0, then that would be a breaking change right?

gwbrown replied:
> Should resolution of _all that includes system indices trigger a deprecation warning?

This does feel odd, but:
1) If it *doesn't* trigger a deprecation warning, then (for example) searches against `_all` that return documents in system indices will have an unannounced breaking change in 8.0, as you note in your second question.
2) As far as I can tell, in the presence of Security, it's impossible to tell in `IndexNameExpressionResolver` if the original query used a wildcard or not, as the original list of indices has by then been replaced. If I'm wrong and there's a good way to get that information, this might be a bit easier to fix.

> If we get all aliases and return system aliases such as .security in 7.x but stop doing so in 8.0, then that would be a breaking change right?

Sort of - if we go ahead with our plan of treating system indices created in 7.x with the previous 7.x behavior, then technically no, but if we upgrade all of the system indices shortly after creation, it'll feel like it to a user.

I'd definitely like to figure out something better to do here.

jaymode replied:
> As far as I can tell, in the presence of Security, it's impossible to tell in IndexNameExpressionResolver if the original query used a wildcard or not, as the original list of indices has by then been replaced

That is 100% correct. I do wonder if we could have security place the original indices in the thread context and have the IndexNameExpressionResolver know that header/context key?

> if we go ahead with our plan of treating system indices created in 7.x with the previous 7.x behavior, then technically no, but if we upgrade all of the system indices shortly after ugprade, it'll feel like it to a user.

I think there may be confusion around treatment of 7.x system indices. The original plan involved every index that was dot-prefixed to be a system index except those created in 7.x or earlier. Once we made the decision to allow hidden indices to be dot-prefixed, that opened up the ability to no longer having to provide this layer of bwc for system indices since we could have user created indices, hidden indices, and system indices that all start with dots after an upgrade. Instead, we allow 7.x dot indices that are not system or hidden to behave normally. 

We're already marking specific indices as system indices in 7.x if a descriptor has been registered for that index to route the read operations to the correct thread pool.

I am wondering if the types removal behavior could fit here:

1. Implicit system index access without the parameter is deprecated in 7.x
2. Explicit system index access without the parameter is deprecated in 7.x
3. System index access with the parameter is not deprecated in 7.x
4. Implicit system index access without the parameter is a bad request in 8.0
5. Explicit system index access without the parameter is a bad request in 8.0
6. System index access with the parameter is deprecated in 8.0+
7. Parameter removed for 9.0 :)

jaymode commented on server/src/main/java/org/elasticsearch/rest/action/cat/AbstractCatAction.java:
```diff
@@ -42,6 +42,11 @@
 
+    @Override
+    public boolean allowSystemIndexAccessByDefault() {
+        return true;
```
why do we allow system by default for cat apis?

gwbrown replied:
This ties into the "Resolving `_all` causes a deprecation warning" issue, and we can't omit them (like we do with hidden indices) in 7.x, as that would break BWC.

jaymode commented on x-pack/docs/en/watcher/managing-watches.asciidoc:
```diff
@@ -28,7 +28,7 @@ For example, the following returns the first 100 watches:
 [source,console]
 --------------------------------------------------
-GET .watches/_search
+GET .watches/_search?allow_system_index_access=true
```
we should open an issue for the watcher docs since we need to stop accessing `.watches` directly

gwbrown replied:
I've added this to the list of tests that need to be adjusted to remove this flag (https://github.com/elastic/elasticsearch/issues/62501).

jaymode commented on x-pack/plugin/data-streams/src/main/java/org/elasticsearch/xpack/datastreams/rest/RestDataStreamsStatsAction.java:
```diff
@@ -29,6 +29,12 @@ public String getName() {
 
+    @Override
+    public boolean allowSystemIndexAccessByDefault() {
+        // Does not actually access system indices, but does resolve `_all`, which can cause confusing deprecation warnings.
```
this feels like a hack to me; I wonder if we can find a way to not need this for items that resolve _all to do certain things :|

gwbrown replied:
It's definitely a hack, and I'd love to not do this. I'll dig in a bit and see if there's a better way to handle this case.

jaymode commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -350,67 +355,80 @@ private void transferDataToEnrichIndex(final String destinationIndexName) {
+        // The ContextPreservingActionListener here is for the purpose of dropping the response headers, as we need this reindex to run
+        // in the security context of the user (rather than Enrich's security context) to ensure that DLS/FLS is correctly applied, but
+        // the reindex needs to access the `.enrich` index, which causes a deprecation warning. Since we drop the response headers,
+        // the deprecation warning is also dropped - but this is a hack and will not work once full protections of system indices are
```
Do we have an issue for this yet?

gwbrown replied:
We do now: https://github.com/elastic/elasticsearch/issues/62505

jaymode commented on x-pack/plugin/src/test/resources/rest-api-spec/test/ml/custom_all_field.yml:
```diff
@@ -148,6 +148,7 @@ setup:
 
   - do:
       search:
+        allow_system_index_access: true
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/transform/qa/single-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/transform/integration/TransformConfigurationIndexIT.java:
```diff
@@ -43,6 +43,7 @@ public void testDeleteConfigurationLeftOver() throws IOException {
             final StringEntity entity = new StringEntity(Strings.toString(builder), ContentType.APPLICATION_JSON);
             Request req = new Request("PUT",
                     TransformInternalIndexConstants.LATEST_INDEX_NAME + "/_doc/" + TransformConfig.documentId(fakeTransformName));
+            req.addParameter("allow_system_index_access", "true");
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/transform/qa/single-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/transform/integration/TransformInternalIndexIT.java:
```diff
@@ -79,14 +76,18 @@ public void testUpdateDeletesOldTransformConfig() throws Exception {
-            RequestOptions.DEFAULT);
-        assertThat(getResponse.isExists(), is(true));
+        Request indexRequest = new Request("PUT", OLD_INDEX + "/_doc/" + TransformConfig.documentId(transformId));
+        indexRequest.addParameter("allow_system_index_access", "true");
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/transform/qa/single-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/transform/integration/TransformRestTestCase.java:
```diff
@@ -480,6 +480,7 @@ public void wipeTransforms() throws IOException {
 
         // the configuration index should be empty
         Request request = new Request("GET", TransformInternalIndexConstants.LATEST_INDEX_NAME + "/_search");
+        request.addParameter("allow_system_index_access", "true");
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/transform/qa/single-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/transform/integration/TransformRobustnessIT.java:
```diff
@@ -117,6 +117,8 @@ private int getNumberOfTransformTasks() throws IOException {
     private void beEvilAndDeleteTheTransformIndex() throws IOException {
-        adminClient().performRequest(new Request("DELETE", TransformInternalIndexConstants.LATEST_INDEX_NAME));
+        final Request deleteRequest = new Request("DELETE", TransformInternalIndexConstants.LATEST_INDEX_NAME);
+        deleteRequest.addParameter("allow_system_index_access", "true");
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/transform/qa/single-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/transform/integration/TransformUsageIT.java:
```diff
@@ -62,6 +62,7 @@ public void testUsage() throws Exception {
                 + ":"
                 + TransformStoredDoc.NAME
         );
+        statsExistsRequest.addParameter("allow_system_index_access", "true");
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/watcher/qa/common/src/main/java/org/elasticsearch/xpack/watcher/WatcherRestTestCase.java:
```diff
@@ -73,6 +73,7 @@ public final void stopWatcher() throws Exception {
 
         Request deleteWatchesIndexRequest = new Request("DELETE", ".watches");
         deleteWatchesIndexRequest.addParameter("ignore_unavailable", "true");
+        deleteWatchesIndexRequest.addParameter("allow_system_index_access", "true");
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/watcher/qa/common/src/main/java/org/elasticsearch/xpack/watcher/WatcherYamlSuiteTestCase.java:
```diff
@@ -105,6 +105,7 @@ public final void stopWatcher() throws Exception {
     private static void deleteWatcherIndices() throws IOException {
         Request deleteWatchesIndexRequest = new Request("DELETE", ".watches");
         deleteWatchesIndexRequest.addParameter("ignore_unavailable", "true");
+        deleteWatchesIndexRequest.addParameter("allow_system_index_access", "true");
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/watcher/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/watcher/ack_watch/10_basic.yml:
```diff
@@ -46,6 +46,7 @@
   - do:
       search:
         rest_total_hits_as_int: true
+        allow_system_index_access: true
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/watcher/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/watcher/activate_watch/10_basic.yml:
```diff
@@ -50,6 +50,7 @@
   - do:
       search:
         rest_total_hits_as_int: true
+        allow_system_index_access: true
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/watcher/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/watcher/delete_watch/10_basic.yml:
```diff
@@ -54,6 +54,7 @@ teardown:
   - do:
       search:
         rest_total_hits_as_int: true
+        allow_system_index_access: true
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/watcher/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/watcher/get_watch/10_basic.yml:
```diff
@@ -49,6 +49,7 @@ teardown:
   - do:
       search:
         rest_total_hits_as_int: true
+        allow_system_index_access: true
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/watcher/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/watcher/put_watch/80_put_get_watch_with_passwords.yml:
```diff
@@ -262,6 +262,7 @@ setup:
   - do:
       search:
         rest_total_hits_as_int: true
+        allow_system_index_access: true
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/plugin/watcher/qa/with-monitoring/src/javaRestTest/java/org/elasticsearch/smoketest/MonitoringWithWatcherRestIT.java:
```diff
@@ -86,8 +88,10 @@ private void assertMonitoringWatchHasBeenOverWritten(String watchId) throws Exce
-            ObjectPath path = ObjectPath.createFromResponse(client().performRequest(new Request("POST", "/.watches/_count")));
+            refreshAllIndices();
+            final Request countRequest = new Request("POST", "/.watches/_count");
+            countRequest.addParameter("allow_system_index_access", "true");
```
add an issue for this test too (or add it to an issue for all tests)

jaymode commented on x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/MlMappingsUpgradeIT.java:
```diff
@@ -163,6 +163,7 @@ private void assertUpgradedConfigMappings() throws Exception {
 
         assertBusy(() -> {
             Request getMappings = new Request("GET", ".ml-config/_mappings");
+            getMappings.addParameter("allow_system_index_access", "true");
```
add an issue for this test too (or add it to an issue for all tests)

jaymode reviewed (commented):
I like it. I left some comments/questions and I think having a issue or issues with a list of tests we need to fix for system index access would be great

gwbrown commented on modules/reindex/src/yamlRestTest/resources/rest-api-spec/test/delete_by_query/80_slices.yml:
```diff
@@ -154,6 +154,7 @@
       indices.refresh: {}
   - do:
       search:
+        allow_system_index_access: true
```
++, I'll open a meta-issue (as I think there are a few too many for individual issues).

gwbrown replied:
I've opened https://github.com/elastic/elasticsearch/issues/62501 with all the places in our REST tests which either 1) use `allow_system_index_access` or 2) swallow the deprecation warning.

AthenaEryma committed:
Use test assumptions instead of setting system properties per review

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Do not emit deprecation warnings on total wildcard queries

AthenaEryma committed:
Remove system index access by default from APIs which no longer need it

AthenaEryma committed:
First cut of security integration for _all handling

AthenaEryma committed:
Add yet another allow_system_index_access flag as necessary

AthenaEryma committed:
Add yet more allow_system_index_access flag as necessary

AthenaEryma committed:
Check if a deprecation warning should be emitted rather than allowing system index access for _all

AthenaEryma committed:
Remove header drop from IndexResolver

AthenaEryma committed:
Set original indices header OUTSIDE of the context preserving listener

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Line lengths

gwbrown commented:
I think this run was in progress when Jenkins died.

@elasticmachine test this please

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Replace allow_system_index_access flags in Java tests with warnings handlers

AthenaEryma committed:
Replace allow_system_index_access flags in yaml tests with warnings handlers

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Remove a couple unnecessary warning checks (they might be back later)

AthenaEryma committed:
The current security index is `.security-7`, not just ``.security`

AthenaEryma committed:
Fix skip:feature block in YAML test

AthenaEryma committed:
Add missing warning handler for current version

AthenaEryma committed:
Undo _all special case handling

AthenaEryma committed:
Remove allow_system_index_access flag

AthenaEryma committed:
More `.security` -> `.security-7` in expected warnings

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Unused import

AthenaEryma committed:
Work around _all resolution in Get Aliases API

AthenaEryma committed:
Specify index in "Test wildcard job id" so we don't get a warning

AthenaEryma committed:
Accept warning in SQL test that might trigger it

AthenaEryma committed:
Allow system index access in APIs we're confident we want it

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Spotless

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Respect system index access header when re-adding deprecation in Get Alias API

AthenaEryma committed:
Handle Get Alias deprecation warning in upgrade test

AthenaEryma committed:
Allow data streams stats action to only resolve to data streams

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Allow deprecation warning on another alias call

AthenaEryma committed:
Allow _cat/aliases to access system indices by default because it's weird otherwise

AthenaEryma committed:
Allow deprecation warning on wildcard Get Aliases call in MlJobIT

AthenaEryma committed:
Remove unnecessary semicolon

gwbrown commented:
I have not yet switched over to suppressing deprecation warnings from Kibana as that's dependent on work still in progress (https://github.com/elastic/kibana/pull/79218), but I think that could be done as follow-up PR if necessary.

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
Fix compilation after merge

gwbrown commented:
Actually, I retract my previous statement about that being doable as a follow-up: Without the `allow_system_index_access` parameter, many tests currently expect the deprecation warning - which won't be present in release builds, so those tests will fail. I think we'll need to make that change first.

AthenaEryma committed:
Remove "feature flag" for deprecation warnings

AthenaEryma committed:
Merge branch 'master' into si/protection-header

jaymode commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -59,19 +63,37 @@
 import java.util.stream.StreamSupport;
 
 public class IndexNameExpressionResolver {
+    public static DeprecationLogger deprecationLogger = DeprecationLogger.getLogger(IndexNameExpressionResolver.class);
```
```suggestion
    private static final DeprecationLogger deprecationLogger = DeprecationLogger.getLogger(IndexNameExpressionResolver.class);
```

jaymode commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -650,6 +693,10 @@ boolean isPatternMatchingAllIndices(Metadata metadata, String[] indicesOrAliases
         return false;
     }
 
+    public boolean isSystemIndexAccessAllowed() {
```
can you document this method?

jaymode commented on server/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -218,7 +219,8 @@ public void dispatchBadRequest(final RestChannel channel, final ThreadContext th
 
-    private void dispatchRequest(RestRequest request, RestChannel channel, RestHandler handler) throws Exception {
+    private void dispatchRequest(RestRequest request, RestChannel channel, RestHandler handler,
+                                 ThreadContext threadContext) throws Exception {
```
technically we don't need to modify this method's signature since the class has the NodeClient, so we could do `client.threadPool().getThreadContext()` to get the ThreadContext and even store the reference in the class constructor.

jaymode commented on server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestRefreshAction.java:
```diff
@@ -51,6 +51,11 @@ public String getName() {
 
+    @Override
+    public boolean allowSystemIndexAccessByDefault() {
+        return true;
```
Do you recall why we need to allow system indices to be refreshed?

gwbrown replied:
I don't think we do - I had intended to remove this.

jaymode commented on server/src/test/java/org/elasticsearch/action/ActionModuleTests.java:
```diff
@@ -146,9 +148,10 @@ public String getName() {
-                    settings.getIndexScopedSettings(), settings.getClusterSettings(), settings.getSettingsFilter(), threadPool,
-                    singletonList(dupsMainAction), null, null, usageService, null);
+            ActionModule actionModule = new ActionModule(settings.getSettings(),
+                new IndexNameExpressionResolver(new ThreadContext(Settings.EMPTY)), settings.getIndexScopedSettings(),
```
```suggestion
                new IndexNameExpressionResolver(threadPool.getThreadContext()), settings.getIndexScopedSettings(),
```

jaymode commented on server/src/test/java/org/elasticsearch/action/ActionModuleTests.java:
```diff
@@ -180,9 +183,10 @@ public void handleRequest(RestRequest request, RestChannel channel, NodeClient c
-                    settings.getIndexScopedSettings(), settings.getClusterSettings(), settings.getSettingsFilter(), threadPool,
-                    singletonList(registersFakeHandler), null, null, usageService, null);
+            ActionModule actionModule = new ActionModule(settings.getSettings(),
+                new IndexNameExpressionResolver(new ThreadContext(Settings.EMPTY)), settings.getIndexScopedSettings(),
```
```suggestion
                new IndexNameExpressionResolver(threadPool.getThreadContext()), settings.getIndexScopedSettings(),
```

jaymode commented on x-pack/docs/en/watcher/managing-watches.asciidoc:
```diff
@@ -33,4 +33,4 @@ GET .watches/_search
 }
 --------------------------------------------------
-// TEST[setup:my_active_watch]
+// TEST[skip:deprecation warning]
```
what is the plan for this documentation?

gwbrown replied:
This is listed in https://github.com/elastic/elasticsearch/issues/62501, to be updated by core/features. While it's not quite a test in the same vein as many of the others, it appears as though it may be an indication that there is a Watcher feature missing (similar to some of the other tests).

jaymode commented on x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/InferenceIngestIT.java:
```diff
@@ -71,8 +70,26 @@ protected Settings restClientSettings() {
+                    // There may not be an index to delete, in which case there's no warning
+                    return false;
+                } else if (warnings.size() > 1) {
+                    logger.warn("too many warnings: {}", warnings);
```
do we need this log statement?

gwbrown replied:
No, thanks for catching. I'll remove.

gwbrown commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/TransportGetAliasesAction.java:
```diff
@@ -85,7 +99,27 @@ protected void masterOperation(Task task, GetAliasesRequest request, ClusterStat
+        return finalResponse;
+    }
+
+    private static void checkSystemIndexAccess(ClusterState state, ImmutableOpenMap<String, List<AliasMetadata>> aliasesMap) {
```
We can do so if it matches a System Index pattern, at least, yes.

gwbrown replied:
I've implemented this such that the alias is only checked if there are no concrete system indices found to warn about.

AthenaEryma committed:
Refresh no longer allow system index access by default

AthenaEryma committed:
Remove unnecessary log line

AthenaEryma committed:
Don't create a new threadcontext, we already have one

AthenaEryma committed:
Don't need to take ThreadContext as a param in dispatchRequest

AthenaEryma committed:
Javadoc for isSystemIndexAccessAllowed

AthenaEryma committed:
Deprecation logger doesn't need to be public

AthenaEryma committed:
Doc fix per review

AthenaEryma committed:
Allow deprecation warning for refreshing all indices

AthenaEryma committed:
Allow warnings on refresh in reindex YAML tests

AthenaEryma committed:
Properly populate client in RestControllerTests now that it's actually used

AthenaEryma committed:
Stash context in RestSubmitAsyncSearchActionTests so the same headers don't get added multiple times

AthenaEryma committed:
Handle more _refresh warnings

AthenaEryma committed:
YAML skip: features: warnings block

AthenaEryma committed:
Address _refresh warnings in yaml tests

AthenaEryma committed:
Warn when accessing aliases with names reserved for system indices

AthenaEryma committed:
Suppress warnings when `X-elastic-product-origin` header is present

AthenaEryma committed:
Merge branch 'master' into si/protection-header

AthenaEryma committed:
EMPTY_MAP -> emptyMap()

AthenaEryma committed:
Invert incorrect header check

AthenaEryma committed:
Add comment about header

AthenaEryma committed:
YAML syntax is THE WORST

AthenaEryma committed:
More specifying _refresh more closely in YAML tests

AthenaEryma committed:
Yet another _refresh call with a warning

AthenaEryma committed:
Fix a couple _refresh calls that got mixed up

AthenaEryma committed:
Even more _refresh index specifying

AthenaEryma committed:
Another _refresh that sparks a warning

jaymode commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/TransportGetAliasesAction.java:
```diff
@@ -25,27 +25,39 @@
+import java.util.stream.Collectors;
 
 public class TransportGetAliasesAction extends TransportMasterNodeReadAction<GetAliasesRequest, GetAliasesResponse> {
+    private static DeprecationLogger deprecationLogger = DeprecationLogger.getLogger(TransportGetAliasesAction.class);
```
```suggestion
    private static final DeprecationLogger deprecationLogger = DeprecationLogger.getLogger(TransportGetAliasesAction.class);
```

jaymode commented on server/src/test/java/org/elasticsearch/rest/RestControllerTests.java:
```diff
@@ -103,8 +107,13 @@ public void setup() {
 
+    @After
+    public void teardown() {
+        client.close();
```
```suggestion
        IOUtils.close(client);
```

This allows us to guard in case the client is null to avoid unnecessary exceptions in case of odd setup failure. The `IOUtils` class I am referencing is from package `org.elasticsearch.core.internal.io`

jaymode commented on test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:
```diff
@@ -798,7 +812,17 @@ private void wipeRollupJobs() throws IOException {
         boolean includeHidden = minimumNodeVersion().onOrAfter(Version.V_7_7_0);
         Request refreshRequest = new Request("POST", "/_refresh");
-        refreshRequest.addParameter("expand_wildcards", "open,closed" + (includeHidden ? ",hidden" : ""));
+        refreshRequest.addParameter("expand_wildcards", "open" + (includeHidden ? ",hidden" : ""));
```
why did closed get dropped here?

gwbrown replied:
Trying to refresh a closed index appears to cause an exception, but that only came up after I switched a test which involves closing an index to use this instead of a direct `POST _refresh` call (to take advantage of the deprecation warning handling). This method is only used in a few tests, and I think specifying `closed` was always wrong, it just never mattered before.

jaymode reviewed (approved):
I left a few minor suggestions and one question. Otherwise LGTM

AthenaEryma committed:
Review feedback

AthenaEryma committed:
Merge branch 'master' into si/protection-header
